src/ImageSharp/Formats/Gif/LzwEncoder.cs

```diff
@@ -11,411 +11,219 @@ namespace SixLabors.ImageSharp.Formats.Gif;
     [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private void FlushPacket(Stream outStream)
+    private void EmitCode(int c, Stream stream)
     {
-        outStream.WriteByte((byte)this.accumulatorCount);
-        outStream.Write(this.accumulators, 0, this.accumulatorCount);
-        this.accumulatorCount = 0;
+        this.cur |= c << this.curShift;
+        this.curShift += this.curCodeSize;
+        this.EmitBytesToBuffer(8, stream);
     }
 
-    /// <summary>
-    /// Output the current code to the stream.
-    /// </summary>
-    /// <param name="code">The code.</param>
-    /// <param name="outs">The stream to write to.</param>
-    private void Output(int code, Stream outs)
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private void EmitBytesToBuffer(int bit_block_size, Stream stream)
     {
-        ref byte accumulatorsRef = ref MemoryMarshal.GetReference(this.accumulators.AsSpan());
-        this.currentAccumulator &= Masks[this.currentBits];
-
-        if (this.currentBits > 0)
-        {
-            this.currentAccumulator |= code << this.currentBits;
-        }
-        else
+        ref byte b = ref MemoryMarshal.GetReference(this.buffer);
+        while (this.curShift >= bit_block_size)
         {
-            this.currentAccumulator = code;
+            Unsafe.Add(ref b, this.position++) = (byte)(this.cur & 0xFF);
+            this.cur >>= 8;
+            this.curShift -= 8;
+            if (this.position == SubBlockLength)
+            {
+                // Finished a sub-block.
+                stream.WriteByte(SubBlockLength);
+                stream.Write(this.buffer);
+                this.position = 0;
+            }
         }
+    }
 
-        this.currentBits += this.bitCount;
+    public void Dispose() => this.bufferOwner.Dispose();
+}
+
+internal ref struct HashLut
+{
+    private readonly IMemoryOwner<int> keysOwner;
+    private readonly IMemoryOwner<int> valuesOwner;
+    private readonly Span<int> keys;
+    private readonly Span<int> values;
+    private nint index;
+
+    public HashLut(MemoryAllocator allocator, int capacity)
+    {
+        this.Capacity = capacity;
+        this.keysOwner = allocator.Allocate<int>(capacity, AllocationOptions.Clean);
+        this.valuesOwner = allocator.Allocate<int>(capacity);
+        this.keys = this.keysOwner.GetSpan();
+        this.keys.Fill(-1);
+        this.values = this.valuesOwner.GetSpan();
+        this.index = 0;
+    }
 
-        while (this.currentBits >= 8)
+    public int Capacity { get; }
+
+    public bool TryGetValue(int key, out int value)
+    {
+        int index = this.keys.IndexOf(key);
+        if (index != -1)
         {
-            this.AddCharacter((byte)(this.currentAccumulator & 0xFF), ref accumulatorsRef, outs);
-            this.currentAccumulator >>= 8;
-            this.currentBits -= 8;
+            ref int v = ref MemoryMarshal.GetReference(this.values);
+            value = Unsafe.Add(ref v, index);
```

Comment:
```suggestion
            value = Unsafe.Add(ref v, (uint)index);
```
To avoid the sign-extending move.

Reply:
I was wondering. Tried a cast to `nint` but didn't consider `uint`